skipruntime-ts/ts/src/skipruntime_api.ts

```diff
@@ -32,6 +32,13 @@ export type MirrorSchema = {
   expected: ColumnSchema[];
   filter?: DBFilter;
 };
+export class MapOptions<K extends TJSON> {
+  constructor(public ranges: [K, K][] | null = null) {}
+}
+export type WithOptions<Params extends Param[], K extends TJSON> = [
+  ...Params,
+  MapOptions<K>?,
+];
```

Comment:
`T?` in a tuple means the element can be omitted

BUT you are right, it also means it can be present with type `undefined`, which is not what I want, unfortunately.

This is indeed ambiguous: when the user writes `(..., undefined)` we don't know whether they meant no `MapOptions` (and the last parameter is `undefined`) or an undefined `MapOptions`.

I've added a last commit which differentiates the two cases in the frontend API and prevents the user from passing an `undefined` `MapOptions`.

There's still a possible ambiguity if the user has a mapper whose last parameter is a `MapOptions`, but why would they?
Maybe I could rely on a function which creates a record with a private unique symbol instead of relying on classes + `instanceof`. Not sure it's worth it for now